mitmproxy/proxy/layers/http/__init__.py

```diff
@@ -708,6 +708,8 @@ def _handle_event(self, event: events.Event):
             yield from self.event_to_child(self.connections[self.context.client], event)
             if self.mode is HTTPMode.upstream:
                 self.context.server.via = server_spec.parse_with_mode(self.context.options.mode)[1]
+        elif isinstance(event, events.Wakeup):
+            yield from self.event_to_child(self.connections[self.context.server], event)
```

Comment:
This always forwards the event to the first server, which may not be the one which requested a wakeup. We need to keep track which stream is the correct one here. Fortunately the same machinery is already in place for `CommandCompleted` events.
I think we can just extend the logic at https://github.com/mitmproxy/mitmproxy/blob/01d67951eed021f4de8dc53f29ae3132a121a348/mitmproxy/proxy/layers/http/__init__.py#L765-L769 to not only register blocking commands but also WakeupRequests, and then forward to the correct layer here similar to what we do for `CommandCompleted` right below.
Let me know if that does not make sense, it's a bit tricky. Happy to expand.